src/DiaryInterface.php

```diff
@@ -25,11 +27,11 @@ public function getBooking($bookingId);
      *
      * @since [*next-version*]
      *
-     * @param mixed $query The query.
+     * @param BookingQueryInterface $query The query instance.
      *
      * @return BookingInterface[] An array of booking instances that match the given query.
      */
-    public function queryBookings($query);
+    public function queryBookings(BookingQueryInterface $query);
```

Comment:
So, as it has been noted in a separate discussion, there is need for a standard (not necessarily backed by interface methods) as to what kind of conditions/expressions are allowed to constitute the `$query` parameter. A set of conditions that can be interpreted locally and definitively needs to be distinguished and documented.

Furthermore, a way to represent "complex" conditions needs to be developed. As an example, there could be the need to provide a nested expression instead of a concrete field value. That expression could require a value of a different entity, one that doesn't yet exist in memory, and thus there is no access to the value itself. The adapter implementation may interpret this as an SQL join, or it may retrieve the secondary entity in the background, or provide  more parameters to a remote API. What matters is that the adapter implementation (what actually does the querying) has all the data it needs in order to perform the query in what _appears_ to the consumer as one operation.